src/btree/bt_cursor.c

```diff
@@ -232,107 +232,197 @@ __cursor_fix_implicit(WT_BTREE *btree, WT_CURSOR_BTREE *cbt)
+        if (cbt->upd_value->type == WT_UPDATE_TOMBSTONE)
+            return (0);
+        *valid = true;
+    }
+    return (0);
+}
+
+/*
+ * __cursor_valid_row --
+ *     Determine if a valid update is being referenced by the cursor for row-store.
+ */
+static int
+__cursor_valid_row(WT_CURSOR_BTREE *cbt, bool *valid, bool check_bounds)
+{
+    WT_DECL_ITEM(key);
     WT_PAGE *page;
     WT_SESSION_IMPL *session;
     WT_UPDATE *upd;
     bool key_out_of_bounds;
 
     *valid = false;
-    key_out_of_bounds = false;
-
-    btree = CUR2BT(cbt);
-    page = cbt->ref->page;
     session = CUR2S(cbt);
     upd = NULL;
+    page = cbt->ref->page;
+    key_out_of_bounds = false;
+
+    /* For all row-store cases we can extract the key from cbt->tmp if the compare value is zero. */
+    if (cbt->compare == 0 && cbt->tmp != NULL)
+        key = cbt->tmp;
 
     /*
-     * We may be pointing to an insert object, and we may have a page with
-     * existing entries.  Insert objects always have associated update
-     * objects (the value).  Any update object may be deleted, or invisible
-     * to us.  In the case of an on-page entry, there is by definition a
-     * value that is visible to us, the original page cell.
-     *
-     * If we find a visible update structure, return our caller a reference
-     * to it because we don't want to repeatedly search for the update, it
-     * might suddenly become invisible (imagine a read-uncommitted session
-     * with another session's aborted insert), and we don't want to handle
-     * that potential error every time we look at the value.
-     *
-     * Unfortunately, the objects we might have and their relationships are
-     * different for the underlying page types.
+     * In the case of row-store, an insert object implies ignoring any page objects, no insert
+     * object can have the same key as an on-page object.
      *
-     * In the case of row-store, an insert object implies ignoring any page
-     * objects, no insert object can have the same key as an on-page object.
      * For row-store:
      *	if there's an insert object:
      *		if there's a visible update:
      *			exact match
      *		else
      *			no exact match
      *	else
-     *		use the on-page object (which may have an associated
-     *		update object that may or may not be visible to us).
+     *		use the on-page object (which may have an associated update object that may or may
+     *		not be visible to us).
+     */
+    WT_RET(__cursor_valid_insert(cbt, key, valid, check_bounds));
+    if (*valid || cbt->ins != NULL)
+        return (0);
+
+    /* The search function doesn't check for empty pages. */
+    if (page->entries == 0)
+        return (0);
+    /*
+     * In case of prepare conflict, the slot might not have a valid value, if the update in the
+     * insert list of a new page scanned is in prepared state.
+     */
+    WT_ASSERT(session, cbt->slot == UINT32_MAX || cbt->slot < page->entries);
```

Comment:
Should this check be before `__cursor_valid_insert` too?

Reply:
I don't think so, in the original implementation it came after the insert list check.